python/extensions/pip.bzl

```
@@ -14,63 +14,135 @@
 
 "pip module extension for use with bzlmod"
 
-load("@rules_python//python/pip_install:pip_repository.bzl", "locked_requirements_label", "pip_repository_attrs", "pip_repository_bzlmod", "use_isolated", "whl_library")
+load("@pythons_hub//:interpreters.bzl", "DEFAULT_PYTHON_VERSION")
+load("@rules_python//python:pip.bzl", "whl_library_alias")
+load(
+    "@rules_python//python/pip_install:pip_repository.bzl",
+    "locked_requirements_label",
+    "pip_hub_repository_bzlmod",
+    "pip_repository_attrs",
+    "pip_repository_bzlmod",
+    "use_isolated",
+    "whl_library",
+)
 load("@rules_python//python/pip_install:requirements_parser.bzl", parse_requirements = "parse")
 
+def _create_pip(module_ctx, pip_attr, whl_map):
+    hub_name = pip_attr.hub_name
+    pip_name = hub_name + "_{}".format(pip_attr.python_version.replace(".", ""))
+    requrements_lock = locked_requirements_label(module_ctx, pip_attr)
+
+    # Parse the requirements file directly in starlark to get the information
+    # needed for the whl_libary declarations below. This is needed to contain
+    # the pip_repository logic to a single module extension.
+    requirements_lock_content = module_ctx.read(requrements_lock)
+    parse_result = parse_requirements(requirements_lock_content)
+    requirements = parse_result.requirements
+    extra_pip_args = pip_attr.extra_pip_args + parse_result.options
+
+    # Create the repository where users load the `requirement` macro. Under bzlmod
+    # this does not create the install_deps() macro.
+    pip_repository_bzlmod(
+        name = pip_name,
+        repo_name = pip_name,
+        requirements_lock = pip_attr.requirements_lock,
+    )
+    if hub_name not in whl_map:
+        whl_map[hub_name] = {}
+
+    for whl_name, requirement_line in requirements:
+        whl_name = _sanitize_name(whl_name)
+        whl_library(
+            name = "%s_%s" % (pip_name, whl_name),
+            requirement = requirement_line,
+            repo = pip_name,
+            repo_prefix = pip_name + "_",
+            annotation = pip_attr.annotations.get(whl_name),
+            python_interpreter = pip_attr.python_interpreter,
+            python_interpreter_target = pip_attr.python_interpreter_target,
+            quiet = pip_attr.quiet,
+            timeout = pip_attr.timeout,
+            isolated = use_isolated(module_ctx, pip_attr),
+            extra_pip_args = extra_pip_args,
+            download_only = pip_attr.download_only,
+            pip_data_exclude = pip_attr.pip_data_exclude,
+            enable_implicit_namespace_pkgs = pip_attr.enable_implicit_namespace_pkgs,
+            environment = pip_attr.environment,
+        )
+
+        if whl_name not in whl_map[hub_name]:
+            whl_map[hub_name][whl_name] = {}
+
+        whl_map[hub_name][whl_name][pip_attr.python_version] = pip_name + "_"
+
 def _pip_impl(module_ctx):
+    pip_hub_map = {}
+    hub_whl_map = {}
```

```suggestion
    # Keeps track of all the hub's whl repos across the different versions.
    # dict[hub, dict[whl, dict[version, str pip]]]
    # Where hub, whl, and pip are the repo names
    hub_whl_map = {}
```

(tbh, this structure strikes me as a bit strange, but I don't fully understand the interaction between things here, so won't press the point).